### DISCUSSION TITLE
GameplayTags and mod support

### DISCUSSION THREAD

**Discussion by ChatouilleGG** [OpenTournament/OpenTournament#219] [Game Features]
I've been diving a bit, to a limited extent, into the GameplayTags system.

It seems the whole thing is designed around having all possible tags pre-defined in project settings (ini files), which are then organized in a complex internal data structure for optimization purposes.

Other than the [GASDocumentation](https://github.com/tranek/GASDocumentation) it is difficult to find information about them, but I also found [this little thread](https://forums.unrealengine.com/development-discussion/c-gameplay-programming/113148-fgameplaytag-at-runtime).

The digging I've done in the code confirms that it is intendedly not possible to add tags after engine startup. It might be possible to modify the system a bit to allow it, but then we need to clean & reconstruct the internal tree which is probably an expensive operation. See __GameplayTagsManager::DoneAddingNativeTags__

System is designed to replicate tags by index in their internal structure instead of by name, for optimization. However this requires server & client to have the **exact same** structure, which in a moddable context is unlikely to happen. It is possible to disable these optimizations though, and force replicate by name. Regardless, the replicated tag still **needs** to be already present on the client machine. See __FGameplayTag::NetSerialize_Packed__

I believe it should be theoretically possible to do something like the following :
- Hack our way into the system to gain the ability to add tags + reconstruct the data structure after engine initialization
- Replicate all server tags to clients when they connect
- Add all missing tags on client machine and rebuild the datastructure
- Disable "fast-replication" (maybe not, if we can ensure we can reproduce the exact same list on server & client then we could benefit from it)

This is without taking into account the work required to let mods add their own tags onto the server, before replicating them to clients. Also applicable to standalone.

**Comment by wail** [comment]
I see two major mod use-cases,

1. Mods that are largely extension of existing behaviors (e.g. new item). More or less see that as a mod that can be delivered within a BP. To me, these should largely be able to work within a defined GameplayTags framework.
2. More extensive mods that involve custom code, more significant extensions possibly with their own binaries. These should (?) be able to define their own GameplayTags architecture.

In either case, 1 or 2, the scope of changes to support new GameplayTags in OT's core branch is literally editing a data file.
